### ISSUE TITLE
BaseURL seems to break DLNA/UPnP

### ISSUE THREAD

**Issue by DarwinAwardWinner** [jellyfin/jellyfin#2377]
**Describe the bug**
I want my Jellyfin server accessible at https://myserver/jellyfin/, so I set the base URL to /jellyfin. However, whenever a base URL is set, UPnP stops working (tested with VLC and `gssdp-discover`). Disabling base URL  (setting it to empty) makes UPnP start working again.

**To Reproduce**

1. Start with base URL disabled.
2. Verify that UPnP is working.
3. Set a base URL and restart Jellyfin.
4. UPnP no longer works (e.g. not visible in VLC's UPnP menu)
5. Unset base URL and restart Jellyfin
6. UPnP works again.

**Expected behavior**
Setting a base URL should not cause UPnP discovery to stop working.

**System (please complete the following information):**
 - OS: Ubuntu 18.04.4 LTS
 - Browser: Firefox
 - Jellyfin Version: 10.4.3-1 and nightly 20200126
 - Installed Plugins: none
 - Reverse proxy: Caddy

**Additional context**
I was able to work around this issue for my use case by disabling Jellyfin's base URL and then configuring my Caddy reverse proxy appropriately, like this:

```
    # Jellyfin server
    proxy /jellyfin/ 127.0.0.1:8096 {
        transparent
        websocket
        without /jellyfin/ # Compensate for no base URL
    }
```

This issue was discussed in Matrix here: https://matrix.to/#/!omTEDrqCdeTFwiOuos:matrix.org/$1580844715840185QCpKn:matrix.org?via=matrix.org&via=bonifacelabs.ca&via=librem.one

**Comment by PrplHaz4** [comment]
This is fixed on master. Will be in 10.5. 

https://github.com/jellyfin/jellyfin/commit/69d76af054dd41cb729edc6d866413aa7e2de622

**Comment by PrplHaz4** [comment]
@dkanada please reopen per https://github.com/jellyfin/jellyfin/issues/1322#issuecomment-586426345

**Comment by mazzystr** [comment]
Sorry guys we don't have it yet....  `Base URL: /jellyfin` is still breaking DLNA.  With `Base URL` set the server is discoverable but content cannot be listed.  Blank out `Base URL` and server is discoverable, content is listable, and content is playable.  

We gotta have `Base URL` working here.  The next step is going to be running without `--net=host`, haproxy ssl offloading, and a broadcast relay between general network and container network.

Version : Jellyfin nightly container / 10.5.0
Clients: VLC android app & Yamaha V-685 amp

**Comment by kylegordon** [comment]
I see #1322 was closed in favour of this issue, but this issue does not mention non-exclusive port bindings as per https://github.com/jellyfin/jellyfin/issues/1322#issuecomment-573026321

Will that aspect of the issue also be covered, or would you rather it was opened as a new issue?

**Comment by JustAMan** [comment]
> would you rather it was opened as a new issue?

It should be a separate issue, as it really _is_ a separate issue :)

**Comment by Bond-009** [comment]
Did someone test this after #2400 was merged?

**Comment by PrplHaz4** [comment]
> 
> 
> Did someone test this after #2400 was merged?

@mazzystr can you confirm the date or commit of the nightly you tested with (or test again with current nightly)?

**Comment by Bond-009** [comment]
I would prefer the second option ;)

**Comment by mazzystr** [comment]
```
$ podman image ls
registry.hub.docker.com/jellyfin/jellyfin     nightly           89ed24cc12ff   8 hours ago
```

I'm getting some descent results.  My previous nightly was from 6 days ago... I deleted that and pulled new nightly.  Initially DLNA failed with my existing config using  `Baseurl /jellyfin`.  I manually edited config.xml Baseurl to /, restarted, DLNA works.  I shutdown, manually edited config.xml Baseurl to /jellyfin, restarted, DLNA works.  

My clients are VLC and Yamaha RX-V685 integrated amplifier.  Not much I can do with the Yamaha other than power down/up and query the server.  That confirms to work.  To be absolutely sure with VLC I forced shutdown of the app, cleared storage/cache, let VLC requery the network and populate the server and directory structure.  Video and audio media play successfully.

I give this 👍 👍

**Comment by mazzystr** [comment]
I've reverted back to `latest` and now my mobile apps are broke.  Sigh... :/

**Comment by mazzystr** [comment]
and I'm back on `nightly 10.5.0` and DLNA still not working.  Mobile apps are good.  Chromecast streams are good.

I bring a tcpdump also...
```
# tcpdump -i eno1 ether broadcast and ether multicast
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eno1, link-type EN10MB (Ethernet), capture size 262144 bytes
16:54:06.544753 IP 192.168.0.169.40885 > 255.255.255.255.netbios-ns: NBT UDP PACKET(137): QUERY; REQUEST; UNICAST
16:54:13.298923 IP 192.168.0.169.40885 > 255.255.255.255.netbios-ns: NBT UDP PACKET(137): QUERY; REQUEST; UNICAST
16:54:16.524819 IP 192.168.0.169.43259 > 255.255.255.255.netbios-ns: NBT UDP PACKET(137): QUERY; REQUEST; UNICAST
^C
3 packets captured
4 packets received by filter
0 packets dropped by kernel
[root@cube ~]# ^C
[root@cube ~]# tcpdump -i eno1 ether broadcast and ether multicast -vvvvv
tcpdump: listening on eno1, link-type EN10MB (Ethernet), capture size 262144 bytes
16:54:22.618230 IP (tos 0x0, ttl 64, id 9687, offset 0, flags [DF], proto UDP (17), length 78)
    192.168.0.169.43259 > 255.255.255.255.netbios-ns: [udp sum ok]
>>> NBT UDP PACKET(137): QUERY; REQUEST; UNICAST
TrnID=0x1B9C
OpCode=0
NmFlags=0x0
Rcode=0
QueryCount=1
AnswerCount=0
AuthorityCount=0
AddressRecCount=0
QuestionRecords:
Name=*               NameType=0x00 (Workstation)
QuestionType=0x20
QuestionClass=0x1


16:54:28.865350 IP (tos 0x0, ttl 64, id 10059, offset 0, flags [DF], proto UDP (17), length 78)
    192.168.0.169.43259 > 255.255.255.255.netbios-ns: [udp sum ok]
>>> NBT UDP PACKET(137): QUERY; REQUEST; UNICAST
TrnID=0x1B9E
OpCode=0
NmFlags=0x0
Rcode=0
QueryCount=1
AnswerCount=0
AuthorityCount=0
AddressRecCount=0
QuestionRecords:
Name=*               NameType=0x00 (Workstation)
QuestionType=0x20
QuestionClass=0x1


^C
2 packets captured
2 packets received by filter
0 packets dropped by kernel
[root@cube ~]#
[root@cube ~]#
[root@cube ~]#
[root@cube ~]# tcpdump -i eno1 ether broadcast and ether multicast -vvvvv
tcpdump: listening on eno1, link-type EN10MB (Ethernet), capture size 262144 bytes
16:55:09.687074 IP (tos 0x0, ttl 64, id 16095, offset 0, flags [DF], proto UDP (17), length 78)
    192.168.0.169.47826 > 255.255.255.255.netbios-ns: [udp sum ok]
>>> NBT UDP PACKET(137): QUERY; REQUEST; UNICAST
TrnID=0x70BB
OpCode=0
NmFlags=0x0
Rcode=0
QueryCount=1
AnswerCount=0
AuthorityCount=0
AddressRecCount=0
QuestionRecords:
Name=*               NameType=0x00 (Workstation)
QuestionType=0x20
QuestionClass=0x1


16:55:15.871804 IP (tos 0x0, ttl 64, id 17143, offset 0, flags [DF], proto UDP (17), length 78)
    192.168.0.169.47826 > 255.255.255.255.netbios-ns: [udp sum ok]
>>> NBT UDP PACKET(137): QUERY; REQUEST; UNICAST
TrnID=0x70BD
OpCode=0
NmFlags=0x0
Rcode=0
QueryCount=1
AnswerCount=0
AuthorityCount=0
AddressRecCount=0
QuestionRecords:
Name=*               NameType=0x00 (Workstation)
QuestionType=0x20
QuestionClass=0x1
```
I feel some data is missing from the tcpdump capture.  I'll keep hacking here.

**Comment by twinkybot** [comment]
Most probably I am affected by this bug as well.
First of all I have a subdomain e.g. jellyfin.example.org. What would my BaseUrl be? I tried "/" but this is removed after restart.
My log says when I try to play via DLNA

```
Error processing request. URL: "http://jelyyfin.example.org/Sessions/ffb30e5672d6215cdd02a581d347b63e/Playing?ItemIds=85458581b2ce6b928d07854964ee6a3d&PlayCommand=PlayNow&StartPositionTicks=2023149430"
System.ArgumentNullException: Value cannot be null. (Parameter 'baseUrl')
   at MediaBrowser.Model.Dlna.StreamInfo.ToUrl(String baseUrl, String accessToken)
   at Emby.Dlna.PlayTo.PlayToController.CreatePlaylistItem(BaseItem item, User user, Int64 startPostionTicks, String mediaSourceId, Nullable`1 audioStreamIndex, Nullable`1 subtitleStreamIndex)
   at Emby.Dlna.PlayTo.PlayToController.SendPlayCommand(PlayRequest command, CancellationToken cancellationToken)
   at Emby.Server.Implementations.Session.SessionManager.SendMessageToSession[T](SessionInfo session, String name, T data, CancellationToken cancellationToken)
   at Emby.Server.Implementations.Session.SessionManager.SendPlayCommand(String controllingSessionId, String sessionId, PlayRequest command, CancellationToken cancellationToken)
   at Emby.Server.Implementations.Services.ServiceExecGeneral.GetTaskResult(Task task)
   at Emby.Server.Implementations.Services.ServiceHandler.ProcessRequestAsync(HttpListenerHost httpHost, IRequest httpReq, HttpResponse httpRes, ILogger logger, CancellationToken cancellationToken)
   at Emby.Server.Implementations.HttpServer.HttpListenerHost.RequestHandler(IHttpRequest httpReq, String urlString, String host, String localPath, CancellationToken cancellationToken)
```
Playing on Laptop or any other device works fine. 

Furthermore when I try to access Jellyfin from the A/V Receiver it shows as DLNA Server "----empty----" which I am not sure what it means. Either that there is no server or that there is no content. I don't trust my device anymore :) but it used to work with Emby with some hiccups but it worked.

**Comment by twinkybot** [comment]
Any insides?
I'd ague as DLNA is a major functionality tat the priority should be higher on this one, no?
Or is the "Known Issues" list sorted ascending in priority (highest at the bottom)?

**Comment by Noir-** [comment]
This is breaking for me as well. I also tried adding `<BaseUrl>/</BaseUrl>` to the config.xml but it seems to get automatically removed.
My client is a LG OLED C9 TV.

**Comment by PacifistRiot** [comment]
Confirmed that base url also breaks DLNA on my LG smart TV using 10.5.5.

**Comment by dkanada** [comment]
The solution is really just to not use a base URL, which should resolve any problems you have been seeing.

**Comment by dkanada** [comment]
Of course, this is definitely something we should fix, but if the setting is unused nothing should be amiss for now.

**Comment by Noir-** [comment]
The thing is that this configuration option isn't even in the settings as far as I can tell. It just renders an installation unusable and you have to debug and manually edit the config to make it work. At the moment, UPNP with Jellyfin is really a mess :confused:

**Comment by twinkybot** [comment]
> The solution is really just to not use a base URL, which should resolve any problems you have been seeing.

No it is not as an empty base URL throws an Exception. What is your proposed solution?

**Comment by dkanada** [comment]
Hmmmm it must be a DLNA only exception then, I'm running without one right now and have zero errors. No proposed solutions though, I don't have a single device that supports DLNA for testing.

> I'd ague as DLNA is a major functionality that the priority should be higher on this one, no?

None of the devs enjoy working on DLNA, so it has a low priority. If someone shows up that uses DLNA and wants to maintain and improve the feature that would change. We can't really do anything about it for now.

**Comment by twinkybot** [comment]
This is sad :( as I need the feature. I used to listen to my music in my living room via my A/V Receiver and now I can only listen to internet radio. Sad.
Like I wrote in reddit. It is a shame that no one wants to write executable specifications as it would have helped enormously to fix this. I want to help but I am utterly lost in the code and I actually don't want to spend so much time to figure out what is wrong :( 
Especially as I do not know how DLNA works in the first place :)

**Comment by Noir-** [comment]
> None of the devs enjoy working on DLNA, so it has a low priority. If someone shows up that uses DLNA and wants to maintain and improve the feature that would change. We can't really do anything about it for now.

What if there would be some people backing this with a bounty?

**Comment by twinkybot** [comment]
> What if there would be some people backing this with a bounty?

Bounties have a draw back. People interest is only to collect the bounty and not for the project, thus they'll most probably hack and do not write maintainable code.

**Comment by Noir-** [comment]
Yeah that's the worst case scenario but luckily it's not about a whole new feature (Isn't it possible to add maintainability to the bounty criteria?). But it sounds like this code isn't maintained anyway. Ideally, the bounty would go to a developer with a certain reputation. You know, I do a lot of things which I don't certainly enjoy but I get money for that :wink: So my hope is that the money helps with prioritization.

If UPNP is so horrible that it's not maintainable maybe it should be removed all together?

**Comment by PrplHaz4** [comment]
I think these two PRs might already make some progress towards fixing this issue - I think the next step is making sure `PublishedServerUrl` is what is sent out for SSDP messages.

@BaronGreenback any idea if this is the case with your new network mgmt implementation?

https://github.com/jellyfin/jellyfin/pull/3033
https://github.com/jellyfin/jellyfin/pull/3257

**Comment by BaronGreenback** [comment]
Perfect timing - just going through Emby.dnla with nullable.

As far as #3257 and ssdp messages go - it only provides the ability to filter them from going out on certain interfaces.

But looking at your exception above on the 16th May - that's being put together in descriptionXmlBuilder.cs where BaseURL is set to the _serverAddress.

This value is calculated from the requested string up to the /dnla/ 

var url = Request.AbsoluteUri;
var serverAddress = url.Substring(0, url.IndexOf("/dlna/", StringComparison.OrdinalIgnoreCase));

**Comment by BaronGreenback** [comment]
Check PR #3481 - may work - may not but be a step closer.

Can't test on my setup at the moment - my current system setup is for other things.

**Comment by BaronGreenback** [comment]
@twinkybot  Done some testing today.  PR #2377  should now work with BaseDir and published-server-url.

If you could test on your system?

**Comment by twinkybot** [comment]
Yes I gladly would but I do not know how. I have Ubuntu and installed Jellyfin via deb. I presume I need to checkout this branch and package it myself to a deb?
Is there a documentation for this?

**Comment by twinkybot** [comment]
So I need to execut build.sh -t native -p ubuntu.amd64

But before that I have to install dotnet like described here: https://docs.microsoft.com/en-us/dotnet/core/install/linux-ubuntu#2004-

**Comment by twinkybot** [comment]
Sorry found the documentation (blush)

**Comment by twinkybot** [comment]
Because I forgot to change the file I had to build it twice and now I am getting an error I do not know what to do:

```
debian/rules:42: recipe for target 'override_dh_auto_build' failed
make[1]: *** [override_dh_auto_build] Error 1
make[1]: Verzeichnis „/home/twinkybot/workspace/jellyfin“ wird verlassen
debian/rules:33: recipe for target 'binary' failed
make: *** [binary] Error 2
dpkg-buildpackage: Fehler: fakeroot debian/rules binary Unterprozess gibt Abbruchstatus 2 zurück
```

**Comment by BaronGreenback** [comment]
Sorry not at all familiar with linux building. However, came across this https://github.com/CESNET/libyang/issues/627 if it's of any help?

**Comment by mazzystr** [comment]
Curiously, I can start/play media via DLNA to my amp (Yamaha RX-V685) with Jellyfin (container jellyfin/jellyfin:nightly-20200709) /jellyfin BaseURL setting if I start it from the web or app.  If I use the amp app and select server source no dlna endpooint is discovered.  Soon as I remove BaseURL setting the amp app discovers the DLNA endpoint is discovered and I can browse and play media.

**Comment by BaronGreenback** [comment]
#2377 hasn't been accepted into the build yet - so you'll need to pull in that code and compile.

It you have already done that change the settings in you logging.default.json file to
{
    "Serilog": {
        "MinimumLevel": {
            "Default": "Warning",
            "Override": {
                "Microsoft": "Warning",
                "System": "Warning",
	"Emby.Dlna": "Debug",
	"Emby.Dlna.Eventing": "Debug"
            }

restart jellyfin - and post the log here.

This will turn on the dlna debugging

**Comment by zzc-tongji** [comment]
I have met the same problem when running Jellyfin 10.5.5 in Docker.

**Comment by zzc-tongji** [comment]
It seems that the problem still exists in 10.6.0 (Docker).

**Comment by BaronGreenback** [comment]
https://github.com/jellyfin/jellyfin/pull/3481

**Comment by BaronGreenback** [comment]
@zzc-tongji  10.6.0 breaks certains other things. Try 10.6.1 and let me know if it works.

**Comment by zzc-tongji** [comment]
@BaronGreenback I have upgraded to 10.6.4 and it seems still doesn't work.

DLAN client as [nPlayer](https://apps.apple.com/us/app/nplayer/id1116905928) says `Premature end of data in tag HTML line 2`.

**Comment by kyz** [comment]
I've been trying to use Jellyfin to play to a Samsung Smart TV.

Installed: Jellyfin 10.6.4-1
OS: Ubuntu 20.04 amd64
Running behind Apache 2 proxy (serving http on port 80 and https on 443, the only ports open on the firewall) 
Base URL is /jellyfin
Local bind address is 127.0.0.1
LAN networks set to 10.0.0.0/255.255.255.0 

I get an exception if local bind address is set to 127.0.0.1:
```
[2020-09-14 12:01:22.277 +00:00] [DBG] Attempting to create PlayToController from location "http://10.0.0.238:7676/smp_16_"
[2020-09-14 12:01:22.277 +00:00] [DBG] Logging session activity from location "http://10.0.0.238:7676/smp_16_"
[2020-09-14 12:01:23.300 +00:00] [DBG] Dlna Device.Start
[2020-09-14 12:01:23.316 +00:00] [DBG] Found matching device profile: "Samsung Smart TV"
[2020-09-14 12:01:23.366 +00:00] [INF] DLNA Session created for "[TV]Samsung LED32" - "UE32F5500"
...
[2020-09-14 12:02:08.399 +00:00] [DBG] "[TV]Samsung LED32" - Received PlayRequest: PlayNow
[2020-09-14 12:02:08.420 +00:00] [DBG] Found matching device profile: "Samsung Smart TV"
[2020-09-14 12:02:08.507 +00:00] [INF] Profile: "Samsung Smart TV", Path: "/video/films/FOO.mp4", isEligibleForDirectPlay: True, isEligibleForDirectStream: True
[2020-09-14 12:02:08.545 +00:00] [ERR] Error processing request. URL: "http://127.0.0.1:8096/jellyfin/Sessions/FOO/Playing?ItemIds=FOO&PlayCommand=PlayNow&StartPositionTicks=0"
System.ArgumentNullException: Value cannot be null. (Parameter 'baseUrl')
   at MediaBrowser.Model.Dlna.StreamInfo.ToUrl(String baseUrl, String accessToken)
   at Emby.Dlna.PlayTo.PlayToController.CreatePlaylistItem(BaseItem item, User user, Int64 startPostionTicks, String mediaSourceId, Nullable`1 audioStreamIndex, Nullable`1 subtitleStreamIndex)
   at Emby.Dlna.PlayTo.PlayToController.SendPlayCommand(PlayRequest command, CancellationToken cancellationToken)
   at Emby.Dlna.PlayTo.PlayToController.SendMessage[T](String name, Guid messageId, T data, CancellationToken cancellationToken)
   at Emby.Server.Implementations.Session.SessionManager.SendMessageToSession[T](SessionInfo session, String name, T data, CancellationToken cancellationToken)
   at Emby.Server.Implementations.Session.SessionManager.SendPlayCommand(String controllingSessionId, String sessionId, PlayRequest command, CancellationToken cancellationToken)
   at Emby.Server.Implementations.Services.ServiceExecGeneral.GetTaskResult(Task task)
   at Emby.Server.Implementations.Services.ServiceHandler.ProcessRequestAsync(HttpListenerHost httpHost, IRequest httpReq, HttpResponse httpRes, ILogger logger, CancellationToken cancellationToken)
   at Emby.Server.Implementations.HttpServer.HttpListenerHost.RequestHandler(IHttpRequest httpReq, String urlString, String host, String localPath, CancellationToken cancellationToken)
```
However, provided the local bind address is left blank, DLNA sending to the TV works both when a base URL is set and when it is blank, and it also works when "LAN networks" are set and when they are blank.

**Comment by BaronGreenback** [comment]
DLNA uses ip address and not urls, so that is why it works.

The system cannot find 127.0.0.1 in the list of ip addresses, hence the null crash.

This whole backend system is currently going through the PR system as i type.
I will run this against the new version to make sure it works.

**Comment by crobibero** [comment]
Should be fixed by https://github.com/jellyfin/jellyfin/pull/4162

**Comment by BaronGreenback** [comment]
Its not been merged with the unstable yet.

**Comment by sam-cross** [comment]
@BaronGreenback I apologise for bringing up a closed issue, but could I ask how I can track which commits _are_ in unstable? Thanks!

**Comment by PrplHaz4** [comment]
> 
> 
> @BaronGreenback I apologise for bringing up a closed issue, but could I ask how I can track which commits _are_ in unstable? Thanks!

#4161 says the fix was merged to master on Sept 27:
>
> Bond-009 merged commit 43add83 into jellyfin:master on Sep 27

At this point you're probably better off waiting for the RC to test, as that has significant network changes that will impact DLNA (for the better)...

To check commit history: https://github.com/jellyfin/jellyfin/commits/master - but beware it's loooong since last release...

**Comment by BaronGreenback** [comment]
I've just checked (and corrected a subsequent bug) - DLNA with baseurl works in network PR 2.

**Comment by cloventt** [comment]
I'm still seeing this issue on 20210616.1-unstable (admin dashboard shows version 10.8.0 linux X64 running inside Docker) and on version 10.7.5 stable.

DLNA works fine without a base URL set. As soon as it is set, DLNA stops working. Unsetting it makes DLNA work again.

**Comment by BaronGreenback** [comment]
What are you setting baseurl to?
